### PR TITLE
Added a function to support data conversion if type is not supported by the taipy.gui.data module

### LINKED ISSUE #457: Tabular Data: support a callback to change unsupported data type to supported one
**What would that feature address**
display unsupported data type

***Description of the ideal solution***
If a dataframe holds a column with unsupported data, we could handle a user callback that would transform that column to a supported type

***Caveats***
performance ?

***Other options***
Raise an error (as of today)

**Acceptance Criteria**
- [ ] Ensure new code is unit tested, and check code coverage is at least 90%
- [ ] Propagate any change on the demos and run all of them to ensure there is no breaking change
- [ ] Ensure any change is well documented

### PR DESCRIPTION
**Pull request by Ujj1225**
Resolves: #457 

So there was this issue where we thought of giving developer option to change unsupported data format to a supported type
For this I have made two functions in class Gui namely "on_invalid_data" and "set_on_invalid_data_callback"

The Second function "set_on_invalid_data_callback" accepts a callback as parameter and uses this function to transform data into supported format. If user doesn't provide a callback, it works as it used to. Nothing would change. 

Besides in data_accessor, I have modified _InvalidDataAccessor. For all methods like get_data, get_col_types, etc. It calls the first function "on_invalid_data" to transform it into a valid data type. If no callback present it will work as it used to work.

### CODE REVIEW

**Comment by Ujj1225** [comment]
@quest-bot loot #457

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: we have our own formatting rules, we kindly request you to respect them

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: I think the callback should not be invoked here.
It should be invoked when evaluating the right `DataAccessor` for the data type.
this way, we can redirect to an existing `DataAccessor` and keep the `InvalidDataAccessor` for invalid data types.

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: on_invalid_data should be handled the same way as on_action

- [definition](https://github.com/Avaiga/taipy/blob/3297952c0e64bd10358fde8e63a3adf6aef4e613/taipy/gui/gui.py#L255)
- [assignment](https://github.com/Avaiga/taipy/blob/3297952c0e64bd10358fde8e63a3adf6aef4e613/taipy/gui/gui.py#L2583)

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
Good work @Ujj1225 
Please, move the invocation of the callback
we need tests

**Comment by Ujj1225** [comment]
hey, @FredLL-Avaiga  
I am so sorry for this formatting issue (turns out my prettier mistakely formatted the code I'll make sure to get it reverted)
So i changed the function on_invalid_data's definition and assignment following the way on_action is defined and assigned. 
Besides I also did run some tests. I havent pushed the file. Its called test_unsupported.py
What scenarios did I test?
- No callback
- Callback that returns None
- Callback that transforms data
- Callback that raises an exception
Its results
image
Do i need to push this file as well? I created it inside gui_specific tests

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: please remove this function

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: could use `_is_function(self.on_invalid_data)`

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: please remove your formatting, it makes reading your changes complicated

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: here we should check the availability of an accessor with the new type and return it.

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

**Comment by Ujj1225** [comment]
hey i wrote a test case for the _get_instance as well 
image

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py]: no need for this line, the user will get the warning

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: _is_function covers None

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: no need for a warning here

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
I don't see your tests ?

**Comment by Ujj1225** [comment]
> I don't see your tests ?

Hey i have pushed test featuring the method __get_instance of data_accessor. This is where we are actually using transformed value and getting the access type for the transformed value

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: you could have register the type through Gui

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py:1]: I think you should test through the public API as much as possible 
ie use get_data() to test __get_instance()

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

**Comment by Ujj1225** [comment]
hey @FredLL-Avaiga 
I am struggling with the tests 
As you suggested i tried to interact with the private get_instance using get_data but seems like get_data expects an object of type _TaipyData
So direct passing of int or str gives me assertion error 
and if i create a mocktaipydata to mimic it then again i get an error saying 
Can't find Data Accessor for type <class 'tests.gui.gui_specific.test_get_instance.MockTaipyData'>

How do i write the tests so it checks my get_instance through public api. I ran into a lot of assertion issues. This is why i tried to access the get_instance directly. 
plus you said i could register the type through gui. Which function does this?

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [reply tests/gui/gui_specific/test_get_instance.py]: gui._get_accessor()._register(<your testing accessor>)

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: gui._get_accessor().get_data(var_name, newvalue, payload)

**Comment by FredLL-Avaiga** [comment]
any news ?
please update your branch

**Comment by FredLL-Avaiga** [comment]
Are you still working on this PR @Ujj1225  ?
If you are, please resolve the conflicts, update your branch and answer the questions
If you are not, please close the PR

**Comment by Ujj1225** [comment]
yes @FredLL-Avaiga  I am working on it I'll resolve the conflicts and update the branch soon!

**Comment by Ujj1225** [comment]
hey @FredLL-Avaiga  I have now swtiched to use of public api _get_data to check for the method _get_instance but still I am struggling with the use of _get_accessor()._register()

**Comment by FredLL-Avaiga** [comment]
your branch still have conflicts

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: make this method private by prefixing it with _

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]:
```
class MyDataAccessor(_DataAccessor):
    @staticmethod
    def get_supported_classes() -> t.List[t.Type]:
        return [int]
  def get_data(
        self, var_name: str, value: t.Any, payload: t.Dict[str, t.Any], data_format: _DataFormat
    ) -> t.Dict[str, t.Any]:
      return {"value": value}
 ...implement
```

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]:
```
data_accessors._register(MyDataAccessor)
result= data_accessors.get_data("var_name", 100, {})
assert result.get("value") == 100
```

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
I began some test implementation
Can you check it out and finish it ?

File: tests/gui/gui_specific/test_get_instance.py
Comment (FredLL-Avaiga) [inline tests/gui/gui_specific/test_get_instance.py]: why don't you use the _register method and the MyDataAccessor class ?

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]
are there still some linter issue ?

**Comment by RymMichaut** [comment]
Congratulations @Ujj1225 on this merged PR
Could you please contact me at rym.michaut@taipy.io for the shipping?

File: tests/gui/gui_specific/test_get_instance.py
Comment (FabienLelaquais) [inline tests/gui/gui_specific/test_get_instance.py]:
```suggestion
    gui._handle_invalid_data = lambda _: None
```

File: tests/gui/gui_specific/test_get_instance.py
Comment (FabienLelaquais) [inline tests/gui/gui_specific/test_get_instance.py]:
```suggestion
    gui.handle_invalid_data = lambda _: 123  # type: ignore
```

**Comment by FabienLelaquais** [comment]
So I have a few remarks after I ultimately got the time to look at this.
First of all well done, @Ujj1225. What you propose does the job, and instructions were followed. And congrats for the tests!
We will integrate this into Taipy BUT only after I'm happy with the proposed API, which I'm not. But hey, nobody could have thought, since I never opened my mouth.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - 

Here are the reasons why I'm reluctant to merge as-is:

- Really, this thing is *not* a callback. In all of Taipy's terminology, a callback invocation is the result of an event that reflects something happening on the front-end. A variable was changed (`on_change`), a button was pressed (`on_action`) and so on.
This is really not the topic here. And again, the original wording of the issue made it confusing. I apologize for that.

- The proposal to create a method of Gui makes things clear... but wrong, in my opinion.
  The functionality is not related to any Gui. A Taipy GUI application with several Gui instances (although we're still wondering what the use case would be) would not really know what to do.

- The role of the user-defined function is to convert data with an unknown type into a typed value that Taipy supports.
  That could be seen as a "type converter," but the naming is confusing since what it does is really convert the data, with the type change being a (desired) side effect.
  So let's call this a "data converter". Actually, that's not so bad, since we can implement whatever we want, right?
  It's a shame there's so little context when the function is invoked (then the conversion could benefit from more information to do things differently), but we have what we have.
  And what we want is to handle values with unknown types. Not so much invalid. Now we get to "unknown data converter".
  And this would be exposed from the *data* module. Finally, "unknown converter" can do the job.
  I would certainly create a global function (`taipy.gui.data.set_unknown_converter()`) that can be invoked with any user-defined function expecting any value (knowing that its type was not recognized by Taipy for serialization) and returning any other value with a type that Taipy does supports.
  A single global variable (similar to current's `Gui.on_invalid_data`) will privately keep this from the user's eyes.

The bottom line is this exposed API:

```
# set_unknown_converter() would be defined in or imported by taipy/gui/data/__init__.py

def my_amazing_converter(value: any) -> Any:
    return ... # Whatever transforms value into something Taipy can use
  
taipy.gui.data.set_unknown_converter(my_amazing_converter)
```
  
It is implemented in a manner very similar to what you have proposed, and I volunteer to refactor all this.
Of course @Ujj1225 the 100 pts are yours.

@FredLL-Avaiga what say you?

**Comment by FabienLelaquais** [comment]
@Ujj1225 are you still there?

**Comment by Ujj1225** [comment]
yes @FabienLelaquais 
I am keen to see on how you are shaping your approach with this one. Waiting patiently. 
Please let me know if i can be of some help

**Comment by FabienLelaquais** [comment]
> yes @FabienLelaquais I am keen to see on how you are shaping your approach with this one. Waiting patiently. Please let me know if i can be of some help

Sure thing. I was for some reason waiting for your approval, but a look at my message clearly shows I should not have.
I'll take care of the rest of this. Thanks again for your help, @Ujj1225! Take care.

File: tests/gui/gui_specific/test_get_instance.py
Comment (FabienLelaquais) [reply tests/gui/gui_specific/test_get_instance.py]: Done

**Comment by FabienLelaquais** [comment]
@Ujj1225, @FredLL-Avaiga 
As promised I handled the refactoring if the original code.
Not much I admit, but  like it better.
Please let me know what you think.
Cheers

File: taipy/gui/gui.py
Comment (FabienLelaquais) [inline taipy/gui/gui.py]: I forgot to remove this

File: tests/gui/data/test_accessors.py
Comment (FabienLelaquais) [inline tests/gui/data/test_accessors.py]:
```suggestion
    # Testing when accessor is registered
```

**Review by FabienLelaquais — CHANGES_REQUESTED** [review]
Details - but.

File: contributors.txt
Comment (FredLL-Avaiga) [inline contributors.txt:1]: how come ? has this file moved ?

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [inline taipy/gui/data/data_accessor.py:146]: is the `type(self.__gui)` needed ?
I would think that `self.__gui.._convert_unsupported_data(value)` would work ?

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py:1152]: this is invoked only when the property is types as TaipyData

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [inline taipy/gui/gui.py]: use `_is_function` instead of `callable`

**Review by FredLL-Avaiga — CHANGES_REQUESTED** [review]

File: taipy/gui/data/data_accessor.py
Comment (FabienLelaquais) [reply taipy/gui/data/data_accessor.py:146]: `_convert_unsupported_data()` is a class method.

File: taipy/gui/gui.py
Comment (FabienLelaquais) [reply taipy/gui/gui.py:1152]: I'll add this... when TaipyData is documented.

File: taipy/gui/gui.py
Comment (FabienLelaquais) [reply taipy/gui/gui.py]: I would recommend not to. 
A class (with `__call__()` is to be a valid value here.
Tell me.

File: contributors.txt
Comment (FabienLelaquais) [reply contributors.txt:1]: This should no have been created indeed

File: taipy/gui/gui.py
Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: I thought `_is_function` was created for this :-(

Comment (FredLL-Avaiga) [reply taipy/gui/gui.py]: but a class without `__call__` would be `callable`

Comment (FabienLelaquais) [reply taipy/gui/gui.py]: Whaat?

Comment (FabienLelaquais) [reply taipy/gui/gui.py]: Oh I see. `MyClass` class is always callable since you can write `MyClass()`... which invokes the constructor.

File: taipy/gui/data/data_accessor.py
Comment (FredLL-Avaiga) [reply taipy/gui/data/data_accessor.py:146]: I get that but it can be called on the instance, no ?

File: contributors.txt
Comment (FredLL-Avaiga) [inline contributors.txt:1]: I don't get why this file is completely changed ?

**Review by FredLL-Avaiga — DISMISSED** [review]

File: contributors.txt
Comment (FabienLelaquais) [reply contributors.txt:1]: My bad - this should definitively go away. We want Ujj1225 in that's all.

Comment (FabienLelaquais) [reply contributors.txt:1]: This file used to be there, then was moved elsewhere. I'll check this out before I merge.
Thanks!

File: taipy/gui/data/data_accessor.py
Comment (FabienLelaquais) [reply taipy/gui/data/data_accessor.py:146]: No it cannot. `<class>.method()` is different from `<instance>.method()`.

**Review by FredLL-Avaiga — APPROVED** [review]

**Review by FabienLelaquais — APPROVED** [review]

**Pull request merged**
